### ISSUE TITLE
helm list failed

### ISSUE THREAD

**Issue by kinderyj** [kubernetes/helm#3816]
helm list
Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%3DTILLER: dial tcp [::1]:8080: getsockopt: connection refused

Output of `helm version`:2.8.2、2.5.0

Output of `kubectl version`:1.6.2

Cloud Provider/Platform (AKS, GKE, Minikube etc.):

**Comment by michelleN** [comment]
Hey @kinderyj. Are you in an RBAC enabled cluster?

This might be a duplicate of #3467 and #3460

**Comment by bacongobbler** [comment]
it looks like https://github.com/kubernetes/helm/pull/3784 might be the fix for this. Can you try [Helm v2.9.0-rc2](https://github.com/kubernetes/helm/releases/tag/v2.9.0-rc2) and see if that fixes things? You'll definitely need to upgrade tiller for the patch to take effect.

**Comment by lrvick** [comment]
Same problem with:

- minikube v0.25.2
- kubernetes v1.9.4
- helm v2.9.0-rc3

To reproduce:
```
$ minikube start
$ helm init
$ helm list
```
output:
```
Starting local Kubernetes v1.9.4 cluster...
Starting VM...
Getting VM IP address...
Moving files into cluster...
Setting up certs...
Connecting to cluster...
Setting up kubeconfig...
Starting cluster components...
Kubectl is now configured to use the cluster.
Loading cached images from config file.
$HELM_HOME has been configured at /home/lrvick/.helm.

Tiller (the Helm server-side component) has been installed into your Kubernetes Cluster.

Please note: by default, Tiller is deployed with an insecure 'allow unauthenticated users' policy.
For more information on securing your installation see: https://docs.helm.sh/using_helm/#securing-your-helm-installation
Happy Helming!

Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp 127.0.0.1:8080: connect: connection refused
```

**Comment by pingthings** [comment]
I have exactly the same issue:
Helm version:
Client: &version.Version{SemVer:"v2.9.0-rc3", GitCommit:"60abcdca41f544caaecb224acbfb92aee11e1f6e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.0-rc3", GitCommit:"60abcdca41f544caaecb224acbfb92aee11e1f6e", GitTreeState:"clean"}

K8s version:
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:55:54Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.0", GitCommit:"fc32d2f3698e36b93322a3465f63a14e9f0eaead", GitTreeState:"clean", BuildDate:"2018-03-26T16:44:10Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

Error from helm list or anything else for that matters.
Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp [::1]:8080: connect: connection refused

This is a brand new clean install of k8s with no cruft.  if someone has a workaround that they could detail it would be greatly appreciated.

**Comment by kinderyj** [comment]
@michelleN
RBAC is not enabled in my cluster.

**Comment by kinderyj** [comment]
@all 
I tried many methods and finally resolve this issue by this workaround:
you should build the image of tiller to add ENV in the image, and  the dockerfile is in below:
FROM gcr.io/kubernetes-helm/tiller:v2.3.1
ENV KUBERNETES_SERVICE_HOST https://YourKubeMasterHostIP:PORT   (I think this line is optional)
ENV KUBERNETES_MASTER YourKubeMasterHostIP:8080 (I think this line  is mandatory)

you can modify this dockerfile to fulling your k8s cluster.

**Comment by bacongobbler** [comment]
So is KUBERNETES_SERVICE_HOST not present in your tiller pod by default? That should be injected by the kubelet IIRC. What cloud provider is this?

**Comment by kinderyj** [comment]
@bacongobbler 
Our k8s cluster is installed in the bare metal server, we don't use any cloud provider.

**Comment by chenww** [comment]
Confirm if add ENV KUBERNETES_MASTER YourKubeMasterHostIP:8080, it will work. This needs be fixed appropriately as we should have no need to specify this env

**Comment by bacongobbler** [comment]
I understand the frustration, however, the assumption in client-go is that the KUBERNETES_MASTER environment variable is injected into your pods, which would be something set up by the kubernetes cluster itself, not by Helm or Tiller. The behaviour is consistent across all the major cloud providers, so it's probably something missing in the bare metal cluster setup guide which breaks this assumption. :/

Because this has been opened (and answered) multiple times in older tickets like #1791 and #2464, I'm going to close this one as a duplicate. Thanks!

**Comment by mattfarina** [comment]
Note, it seems there are cases where that env var doesn't get set in the pod. For example see https://github.com/kubernetes/kubernetes/issues/40973. Could that be why people are asking here?

**Comment by bacongobbler** [comment]
I think @adamreese also pointed out he saw this issue as well. Seems like it's a separate issue than #2464 so I'm re-opening.

**Comment by jharshman** [comment]
Seeing this issue in GKE as well.

**Comment by muratsplat** [comment]
Yes I have same issue on GKE :| 
@chenww  Where should define ENV `KUBERNETES_MASTER` on ? on cluster on or local machine which helm calling on ?

**Comment by bacongobbler** [comment]
For those seeing this issue, please try 

```
helm init --service-account default
```

and see if that works. See #3990 for more context.

**Comment by muratsplat** [comment]
Thank  you  for the suggestion @bacongobbler but it doest work again..

Helm is more complex more than take care about kubernetes for me . I prefer to  keep simple my case.

**Comment by jharshman** [comment]
@chenww They are suggesting that `KUBERNETES_MASTER` needs to be set on the tiller pod.  However, this doesn't appear to be the case at all.  We have a K8S cluster & helm deployment that works just fine in AWS and that env variable is nowhere to be found.  Only difference between the working deployment of Helm/Tiller and the not working one is the version.

I'm going to roll back to an earlier version of Helm in my GKE cluster later and see if that fixes the issue.

**Comment by jharshman** [comment]
@muratsplat Works fine with version 2.8.2 this looks like a bug that I hope the developers are aware of.
And for those who are interested here is the list of KUBE related env variables in Tiller.  There is no `KUBERNETES_MASTER` env variable.

```
KUBE_DNS_PORT_53_UDP_PROTO
KUBERNETES_DASHBOARD_PORT_443_TCP
KUBE_DNS_SERVICE_PORT
KUBE_DNS_PORT
KUBERNETES_DASHBOARD_SERVICE_HOST
KUBE_DNS_PORT_53_UDP_PORT
KUBE_DNS_PORT_53_TCP_ADDR
KUBERNETES_PORT_443_TCP_PORT
KUBERNETES_PORT
KUBE_DNS_SERVICE_PORT_DNS
KUBE_DNS_PORT_53_TCP
KUBERNETES_PORT_443_TCP
KUBERNETES_DASHBOARD_PORT_443_TCP_PROTO
KUBERNETES_SERVICE_HOST
KUBERNETES_DASHBOARD_PORT_443_TCP_ADDR
KUBE_DNS_SERVICE_PORT_DNS_TCP
KUBE_DNS_PORT_53_UDP_ADDR
KUBERNETES_DASHBOARD_PORT_443_TCP_PORT
KUBERNETES_SERVICE_PORT
KUBERNETES_PORT_443_TCP_ADDR
KUBERNETES_DASHBOARD_SERVICE_PORT
KUBE_DNS_PORT_53_UDP
KUBE_DNS_PORT_53_TCP_PORT
KUBERNETES_SERVICE_PORT_HTTPS
KUBERNETES_DASHBOARD_PORT
KUBE_DNS_SERVICE_HOST
KUBE_DNS_PORT_53_TCP_PROTO
KUBERNETES_PORT_443_TCP_PROTO
```

**Comment by bacongobbler** [comment]
The original issue is probably something else entirely, so we were trying to do some diagnostic testing. v2.9.0 is a differnt bug altogether that just unfortunately results in the same error. See #3985 for some background context on the bug, the fix, and the workaround (as I described above).

@muratsplat I'm not sure I follow. What do you mean by "it doesn't work again"? How did you apply the workaround? You need to run `kubectl -n kube-system delete deploy tiller-deploy` if that's the issue.

**Comment by jharshman** [comment]
@bacongobbler where is this `KUBERNETES_MASTER` envvar coming from? I can't find any references to it in the source code.

**Comment by muratsplat** [comment]
Hello @bacongobbler,

If I can have a time to try to again use helm, when it is the time, I will sharing one step one on here.

**Comment by somedayiamold** [comment]
adding ENV KUBERNETES_MASTER works for me
I am using v2.9.0

**Comment by vladi-dev** [comment]
Facing the same issue in gcloud kubernetes engine.

**Comment by danvaida** [comment]
[This](https://github.com/kubernetes/helm/issues/3460#issuecomment-385817429) fixed it for me on v2.9.0

**Comment by bacongobbler** [comment]
This has been fixed in Helm v2.9.1. Thanks everyone!

https://github.com/kubernetes/helm/releases/tag/v2.9.1

**Comment by kundanatre** [comment]
This is still a issue in 2.9.1 as well.

```
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
```

```
[debug] Created tunnel using local port: '41812'

[debug] SERVER: "127.0.0.1:41812"

Error: Get http://localhost:8080/api/v1/namespaces/kube-system/configmaps?labelSelector=OWNER%!D(MISSING)TILLER: dial tcp 127.0.0.1:8080: connect: connection refused
```
